src/protocol/Connection.cpp

```diff
@@ -222,6 +231,91 @@ void ConnectionPrivate::socketDisconnected()
 
 void ConnectionPrivate::socketReadable()
 {
+    if (!handshakeDone) {
+        qint64 available = socket->bytesAvailable();
+
+        if (direction == Connection::ClientSide && available >= 1) {
+            // Expecting a single byte in response with the chosen version
+            uchar version = ProtocolVersionFailed;
+            if (socket->read(reinterpret_cast<char*>(&version), 1) < 1) {
+                qDebug() << "Connection socket error" << socket->error() << "during read:" << socket->errorString();
+                socket->abort();
+                return;
+            }
+
+            handshakeDone = true;
+            if (version == 0) {
+                qDebug() << "Server in outbound connection is using the version 1.0 protocol";
+                emit q->oldVersionNegotiated(socket);
+                q->close();
+                return;
+            } else if (version != ProtocolVersion) {
+                qDebug() << "Version negotiation failed on outbound connection";
+                emit q->versionNegotiationFailed();
+                socket->abort();
+                return;
+            } else
+                emit q->ready();
+        } else if (direction == Connection::ServerSide && available >= 3) {
+            // Expecting at least 3 bytes
+            uchar intro[3] = { 0 };
+            qint64 re = socket->peek(reinterpret_cast<char*>(intro), sizeof(intro));
+            if (re < (int)sizeof(intro)) {
```

Comment:
I think no: I assume there are all sorts of reasons for a read/peek to fail. I'm mostly only concerned if peek succeeds and then read fails, or if we get some data but less than bytesAvailable promised.